### PR TITLE
Resolve code breakage due to PyTorch scalar introduction

### PR DESCRIPTION
**Pull request by neerajprad**
Addresses #771 

Some minor changes to bring Pyro `dev` in sync with PyTorch master. Tests will fail unless we update the CI wheels, though I have tested locally that all tests pass.

@fritzo - let me know if this looks reasonable. I'll update the CI wheels and re-run the tests.

### CODE REVIEW

File: tutorial/source/bayesian_regression.ipynb [hunk 1]
```diff
@@ -123,7 +123,7 @@
     "    y_data = data[:, -1]\n",
     "    for j in range(num_iterations):\n",
     "        # run the model forward on the data\n",
-    "        y_pred = regression_model(x_data)\n",
+    "        y_pred = regression_model(x_data).squeeze(-1)\n",
```
Comment (neerajprad) [inline tutorial/source/bayesian_regression.ipynb]: @jpchen - `y_pred` returns a tensor of size `100x1` on PyTorch master now, whereas `y_data` has size `100` and hence it errors when we compute the loss. Also, removed the other `squeeze` operations without which things seem to run fine.

File: pyro/infer/tracegraph_elbo.py [hunk 2]
```diff
@@ -155,7 +155,8 @@ def _compute_elbo_non_reparam(guide_trace, guide_vec_md_nodes,  #
         if log_pdf_key == 'log_pdf':
             score_function_term = score_function_term.sum()
         if use_nn_baseline or use_decaying_avg_baseline or use_baseline_value:
-            if downstream_cost.size() != baseline.size():
+            if (downstream_cost.dim() == 0 and baseline.dim() > 1) or \
```
Comment (neerajprad) [inline pyro/infer/tracegraph_elbo.py]: @fritzo, @martinjankowiak - This works fine now, but please feel free to modify this logic later. The issue is that `downstream_cost` can be a dim 0 tensor now whereas the baseline will always have dim > 0 as the output from a nn module can't be dim 0. I tried a couple of other approaches, but this seemed like the cleanest way to get all the tests working.

File: pyro/infer/abstract_infer.py [hunk 3]
```diff
@@ -56,14 +56,16 @@ def _dist_and_values(self, *args, **kwargs):
...
-                logits[ix] = util.log_sum_exp(torch.stack([logits[ix], logit]).squeeze())
+                logits[ix] = util.log_sum_exp(torch.stack([logits[ix], logit]))
 
         logits = torch.stack(logits).squeeze()
         logits -= util.log_sum_exp(logits)
         if not isinstance(logits, torch.autograd.Variable):
             logits = Variable(logits)
         logits = logits - util.log_sum_exp(logits)
-
+        # Categorical distribution must have at least one event dimension
```
Comment (fritzo) [inline pyro/infer/abstract_infer.py]: Would it be cleaner to replace the `.squeeze()` above with a `.view()`?
```diff
- logits = torch.stack(logits).squeeze()
+ logits = torch.stack(logits).contiguous().view(-1)
```

File: tutorial/source/bayesian_regression.ipynb (hunk 1 above)
Comment (jpchen) [reply tutorial/source/bayesian_regression.ipynb]: this looks fine, can you please change this in the example as well? and test with python3 also if you havent, since there was a squeeze bug in pytorch when doing `backward()`.

File: pyro/util.py [hunk 4]
```diff
@@ -202,16 +202,9 @@ def is_inf(x):
...
-        vecs = vecs.view(1, -1)
-    _, idx = torch.max(vecs, 1)
-    max_score = torch.index_select(vecs, 1, idx.view(-1))
-    ret = max_score + torch.log(torch.sum(torch.exp(vecs - max_score.expand_as(vecs))))
-    if n == 1:
-        return ret.view(-1)
-    return ret
+def log_sum_exp(tensor):
+    max_val = tensor.max(dim=-1)[0]
+    return max_val + (tensor - max_val).exp().sum(dim=-1).log()
```
Comment (jpchen) [inline pyro/util.py]: +1. this kind of code simplification is great.

Comment (fritzo) [inline pyro/util.py]: Don't you need to `.unsqueeze(-1)` the `max_val` to broadcast correctly?
```py
return max_val + (tensor - max_val.unsqueeze(-1)).exp().sum(dim=-1).log()
```

Comment (neerajprad) [reply pyro/util.py]: Completely my bad, will change.

File: pyro/infer/abstract_infer.py (hunk 3 above)
Comment (neerajprad) [reply pyro/infer/abstract_infer.py]: Done

**Comment by neerajprad** [comment]
Wheels are updated, so this should be good to merge if all tests pass.

**Comment by neerajprad** [comment]
> Oh I forgot, could you update the version hash in the README.md?

I forgot about that too! Will do.

File: README.md [hunk 5]
```diff
@@ -53,7 +53,7 @@ pip install .
 
 For recent features you can install Pyro from source.
 
-First install a recent PyTorch, currently PyTorch commit `ca5071d07`.
+First install a recent PyTorch, currently PyTorch commit `853dba8`.
```
Comment (fritzo) [inline README.md]: Sorry, you also need to replace the commit below in `git checkout`. Feel free to reword so the commit only appears once.

**Comment by neerajprad** [comment]
This should be good to merge now.
